### ISSUE TITLE
Problem in Curator client to renew Zookeeper IP after pod restart

### ISSUE THREAD

**Issue by RaulGracia** [pravega/pravega#3651]
**Problem description**
I have reproduced the problem described in #3477, as it becoming very difficult to get enough information from longevity runs. This issue intends to better describe the root cause for the unexpected Controller restart upon a Zookeeper session expiration (also changing the `connectString`), which should be actually a graceful restart.

_Scenario_: To reproduce the issue, I created a 3 VM PKS cluster. I deployed the Operators, Zookeeper (1 instance) and Pravega (3BK, 1Controller, 1SSS). The versions for Pravega are the ones in the PR branch #3481.

_Verification_: I verified that the cluster is working properly by executing a couple of short benchmarks, just for being sure that the cluster is healthy before reproducing the issue.

_Issue reproduction_: I tried to reproduce the Controller issue for gracefully restarting upon a ZK session expiration by manually killing the ZK pod (`kubectl delete pod zookeeper-0`).

_Case 1: using as zookeeperUri the client endpoint_: In this first experiment, I used as `zookeeperUri: zookeeper-client:2181` in the `pravega.yml` manifest. This is how other services are supposed to discover the ZK service instances, and this is the way longevity runs are being executed. In this case, killing the ZK pod leads to a Controller pod restart, as we are seeing in longevity runs (which is good, as it is reproducible). In this controlled scenario, using `kubectl logs -f controller-xxx > file.log` provides us with more logs than the ones collected by fluentbit. Here is the summary of what happens:
1) The ZK session expiration is detected with a different `connectString`:
```
2019-04-17 16:29:13,741 394228 [ControllerServiceMain-EventThread] WARN  i.p.c.s.client.StoreClientFactory - Input argument for new ZooKeeper client (172.24.57.9:2181, 10000, false) changed with respect to existing client (zookeeper-client:2181, 10000, false).
```

2) The `ControllerServiceMain` realizes about this, and starts the graceful restart:
```
2019-04-17 16:29:13,743 394230 [ControllerServiceMain] INFO  i.p.c.server.ControllerServiceMain - ZK session expired
2019-04-17 16:29:13,743 394230 [ControllerServiceMain] INFO  i.p.c.server.ControllerServiceMain - Stopping ControllerServiceStarter
```

3) After all the internal modules in the Controller have been shutdown (e.g., REST API server, EventProcessors, etc.), `ControllerServiceMain` finishes the restart:
```
2019-04-17 16:29:59,867 440354 [ControllerServiceStarter STOPPING] INFO  i.p.c.s.ControllerServiceStarter - Finishing controller service shutDown
2019-04-17 16:29:59,867 440354 [ControllerServiceMain] INFO  i.p.c.server.ControllerServiceMain - calling close on store client
```

4) As expected, `ControllerServiceMain` creates a new ZK client to restart the Controller internally:
```
2019-04-17 16:29:59,867 440354 [ControllerServiceMain] INFO  i.p.c.server.ControllerServiceMain - Creating store client
2019-04-17 16:29:59,868 440355 [ControllerServiceMain] WARN  o.a.curator.CuratorZookeeperClient - session timeout [10000] is less than connection timeout [15000]
2019-04-17 16:29:59,868 440355 [ControllerServiceMain] INFO  o.a.c.f.imps.CuratorFrameworkImpl - Starting
2019-04-17 16:29:59,870 440357 [ControllerServiceMain] INFO  org.apache.zookeeper.ZooKeeper - Initiating client connection, connectString=zookeeper-client:2181 sessionTimeout=10000 watcher=org.apache.curator.ConnectionState@2858cf2
2019-04-17 16:29:59,871 440358 [ControllerServiceMain] INFO  o.apache.zookeeper.ClientCnxnSocket - jute.maxbuffer value is 4194304 Bytes
2019-04-17 16:29:59,940 440427 [ControllerServiceMain-SendThread(zookeeper-client:2181)] INFO  org.apache.zookeeper.ClientCnxn - Opening socket connection to server zookeeper-client/10.100.200.30:2181. Will not attempt to authenticate using SASL (unknown error)
2019-04-17 16:29:59,942 440429 [ControllerServiceMain] INFO  o.a.c.f.imps.CuratorFrameworkImpl - Default schema
2019-04-17 16:29:59,943 440430 [ControllerServiceMain] INFO  i.p.c.server.ControllerServiceMain - Awaiting ZK client connection to ZK server
2019-04-17 16:30:09,949 450436 [ControllerServiceMain-SendThread(zookeeper-client:2181)] WARN  org.apache.zookeeper.ClientCnxn - Client session timed out, have not heard from server in 10076ms for sessionid 0x0
2019-04-17 16:30:09,949 450436 [ControllerServiceMain-SendThread(zookeeper-client:2181)] INFO  org.apache.zookeeper.ClientCnxn - Client session timed out, have not heard from server in 10076ms for sessionid 0x0, closing socket connection and attempting reconnect
2019-04-17 16:30:11,050 451537 [ControllerServiceMain-SendThread(zookeeper-client:2181)] INFO  org.apache.zookeeper.ClientCnxn - Opening socket connection to server zookeeper-client/10.100.200.30:2181. Will not attempt to authenticate using SASL (unknown error)
```

5) At this moment, 2 things happen: i) The Controller continues the restart:
```
2019-04-17 16:30:18,129 458616 [ControllerServiceStarter STARTING] INFO  i.p.c.s.ControllerServiceStarter - Initiating controller service startUp
2019-04-17 16:30:18,130 458617 [ControllerServiceStarter STARTING] INFO  i.p.c.s.ControllerServiceStarter - Event processors enabled = true
2019-04-17 16:30:18,130 458617 [ControllerServiceStarter STARTING] INFO  i.p.c.s.ControllerServiceStarter - Cluster listener enabled = true
2019-04-17 16:30:18,130 458617 [ControllerServiceStarter STARTING] INFO  i.p.c.s.ControllerServiceStarter -     Host monitor enabled = true
2019-04-17 16:30:18,130 458617 [ControllerServiceStarter STARTING] INFO  i.p.c.s.ControllerServiceStarter -      gRPC server enabled = true
2019-04-17 16:30:18,130 458617 [ControllerServiceStarter STARTING] INFO  i.p.c.s.ControllerServiceStarter -      REST server enabled = true
```
and ii) the Controller receives a `SIGTERM`, given that the `shutdownHook` added to the `Main` class is invoked:
```
2019-04-17 16:30:13,040 453527 [Thread-2] INFO  io.pravega.controller.server.Main - Shutdown hook memory usage dump: Heap memory usage: init = 536870912(524288K) used = 46706264(45611K) committed = 518979584(506816K) max = 518979584(506816K), non heap memory usage init = 2555904(2496K) used = 67733704(66146K) committed = 71991296(70304K) max = -1(-1K)
```

6) Receiving this unexpected `SIGTERM` is what makes the Controller to restart ungracefully, and then see an increment in the number of pod restarts:
```
kubectlget pods --all-namespaces
NAMESPACE     NAME                                                    READY   STATUS      RESTARTS   AGE
default       benchmark-pod                                           1/1     Running     0          5h51m
default       pravega-bookie-0                                        1/1     Running     0          102m
default       pravega-bookie-1                                        1/1     Running     0          102m
default       pravega-bookie-2                                        1/1     Running     0          102m
default       pravega-operator-5f94fc5c88-jdk9w                       1/1     Running     0          4h59m
default       pravega-pravega-controller-568796cc77-mc5sp             1/1     Running     1          102m
default       pravega-pravega-segmentstore-0                          1/1     Running     0          102m
```

_Case 2: using as zookeeperUri a headless endpoint_: As a test, I changed the `zookeeperUri: zookeeper-0.zookeeper-headless.default.svc.cluster.local:2181`. This is basically pointing to one of the instances (we only have one, actually). Despite this is not the regular way to proceed, I did this attempt just to understand the root cause for this issue. I cleaned up the cluster, executed the same warm-up benchmarks, and killed the ZK pod as before.

1) The situation is very similar to the previous one. We get the ZK session expiration with the different `connectString`:
```
2019-04-17 14:50:25,548 3111148 [ControllerServiceMain-EventThread] WARN  i.p.c.s.client.StoreClientFactory - Input argument for new ZooKeeper client (172.24.57.5:2181, 10000, false) changed with respect to existing client (zookeeper-0.zookeeper-headless.default.svc.cluster.local:2181, 10000, false).
```

2) The graceful restart completes successfully:
```
2019-04-17 14:50:25,561 3111161 [ControllerServiceMain] INFO  i.p.c.server.ControllerServiceMain - ZK session expired
2019-04-17 14:50:25,561 3111161 [ControllerServiceMain] INFO  i.p.c.server.ControllerServiceMain - Stopping ControllerServiceStarter
...
2019-04-17 14:51:12,467 3158067 [ControllerServiceStarter STOPPING] INFO  i.p.c.s.ControllerServiceStarter - Closing store
2019-04-17 14:51:12,470 3158070 [ControllerServiceStarter STOPPING] INFO  i.p.c.s.ControllerServiceStarter - Finishing controller service shutDown
2019-04-17 14:51:12,470 3158070 [ControllerServiceMain] INFO  i.p.c.server.ControllerServiceMain - calling close on store client
```

3) Then, `ControllerServiceMain` creates a new ZK client and starts again the internal modules of the Controller:
```
2019-04-17 14:51:12,470 3158070 [ControllerServiceMain] INFO  i.p.c.server.ControllerServiceMain - Creating store client
2019-04-17 14:51:12,470 3158070 [ControllerServiceMain] WARN  o.a.curator.CuratorZookeeperClient - session timeout [10000] is less than connection timeout [15000]
2019-04-17 14:51:12,470 3158070 [ControllerServiceMain] INFO  o.a.c.f.imps.CuratorFrameworkImpl - Starting
2019-04-17 14:51:12,472 3158072 [ControllerServiceMain] INFO  org.apache.zookeeper.ZooKeeper - Initiating client connection, connectString=zookeeper-0.zookeeper-headless.default.svc.cluster.local:2181 sessionTimeout=10000 watcher=org.apache.curator.ConnectionState@384c8b1a
...
2019-04-17 14:51:12,488 3158088 [ControllerServiceMain] INFO  i.p.c.server.ControllerServiceMain - Awaiting ZK session expiry or termination trigger for ControllerServiceMain
2019-04-17 14:51:12,488 3158088 [ControllerServiceMain] INFO  i.p.c.server.ControllerServiceMain - Starting controller services
2019-04-17 14:51:12,490 3158090 [ControllerServiceMain] INFO  i.p.c.server.ControllerServiceMain - Awaiting controller services start
2019-04-17 14:51:12,490 3158090 [ControllerServiceStarter STARTING] INFO  i.p.c.s.ControllerServiceStarter - Initiating controller service startUp
```

4) But in this case, there is no log for `shutdownHook`. That is, the "same" Controller process is correctly restarted (no pod restarts appear):
```
kubectl get pods
NAME                                                    READY   STATUS    RESTARTS   AGE
benchmark-pod                                           1/1     Running   0          3h21m
pravega-bookie-0                                        1/1     Running   0          96m
pravega-bookie-1                                        1/1     Running   1          96m
pravega-bookie-2                                        1/1     Running   0          96m
pravega-operator-5f94fc5c88-jdk9w                       1/1     Running   0          149m
pravega-pravega-controller-568796cc77-4d7bx             1/1     Running   0          96m
pravega-pravega-segmentstore-0                          1/1     Running   0          96m
wandering-emu-nfs-client-provisioner-5b8b6dd9b8-w668k   1/1     Running   0          100m
zookeeper-0                                             1/1     Running   0          44m
zookeeper-operator-5cf75dcb56-dq6kj                     1/1     Running   0          103m
```

_SUMMARY_: First, with the changes of this PR it seems that the restart logic for the Controller looks good. Second, by comparing both scenarios, it seems that what differentiates a successful graceful restart from a non successful one in the Controller is the `zookeeperUri`. This leads me to think of two main components that may be involved in the problem: the Kubernetes DNS service and Curator: is Curator correctly getting the endpoints for a service URL in K8? Is K8 correctly updating the endpoints for a service when a pod restarts? These are some hypothesis, but your feedback here would be greatly appreciated.

_BONUS TRACK:_ After restarting the ZK pod, the Segment Store was unable to get a proper connection to the new pod. Irrespective of the `zookeeperUri`, the Segment Store got stuck trying to connect to the old IP address (note that the Segment Store also uses Curator):

```
(Case 1: zookeeperUri: zookeeper-client:2181)
--------------------------------------------------------
2019-04-17 16:28:59,948 321819 [core-1] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [0, 1, 2, 3], Current = [0, 1, 2, 3], PendingTasks = [], ToStart = [], ToStop = [].
2019-04-17 16:29:01,446 323317 [main-SendThread(zookeeper-client:2181)] WARN  org.apache.zookeeper.ClientCnxn - Client session timed out, have not heard from server in 6666ms for sessionid 0x1000174bd5c0021
2019-04-17 16:29:01,447 323318 [main-SendThread(zookeeper-client:2181)] INFO  org.apache.zookeeper.ClientCnxn - Client session timed out, have not heard from server in 6666ms for sessionid 0x1000174bd5c0021, closing socket connection and attempting reconnect
2019-04-17 16:29:01,548 323419 [main-EventThread] INFO  org.apache.curator.ConnectionState - Connection string changed to: zookeeper-client:2181
2019-04-17 16:29:01,550 323421 [main-EventThread] INFO  o.a.c.f.s.ConnectionStateManager - State change: SUSPENDED
2019-04-17 16:29:01,550 323421 [Curator-Framework-0] INFO  org.apache.curator.ConnectionState - Connection string changed to: zookeeper-client:2181
2019-04-17 16:29:01,651 323522 [main-SendThread(zookeeper-client:2181)] WARN  org.apache.zookeeper.ClientCnxn - Client session timed out, have not heard from server in 6669ms for sessionid 0x1000174bd5c0022
2019-04-17 16:29:01,651 323522 [main-SendThread(zookeeper-client:2181)] INFO  org.apache.zookeeper.ClientCnxn - Client session timed out, have not heard from server in 6669ms for sessionid 0x1000174bd5c0022, closing socket connection and attempting reconnect
2019-04-17 16:29:01,752 323623 [main-EventThread] INFO  o.a.b.z.ZooKeeperWatcherBase - ZooKeeper client is disconnected from zookeeper now, but it is OK unless we received EXPIRED event.
2019-04-17 16:29:02,551 324422 [Curator-Framework-0] INFO  org.apache.curator.ConnectionState - Connection string changed to: zookeeper-client:2181
2019-04-17 16:29:02,873 324744 [main-SendThread(zookeeper-client:2181)] INFO  org.apache.zookeeper.ClientCnxn - Opening socket connection to server zookeeper-client/10.100.200.30:2181. Will not attempt to authenticate using SASL (unknown error)
2019-04-17 16:29:02,942 324813 [main-SendThread(zookeeper-client:2181)] INFO  org.apache.zookeeper.ClientCnxn - Opening socket connection to server zookeeper-client/10.100.200.30:2181. Will not attempt to authenticate using SASL (unknown error)
...
2019-04-17 16:30:23,596 405467 [main-SendThread(172.24.57.9:2181)] INFO  org.apache.zookeeper.ClientCnxn - Opening socket connection to server 172.24.57.9/172.24.57.9:2181. Will not attempt to authenticate using SASL (unknown error)
2019-04-17 16:30:26,060 407931 [BookKeeperClientScheduler-OrderedScheduler-0-0] INFO  o.a.b.net.NetworkTopologyImpl - Adding a new node: /default-rack/pravega-bookie-0.pravega-bookie-headless.default.svc.cluster.local:3181
2019-04-17 16:30:26,668 408539 [main-SendThread(172.24.57.9:2181)] WARN  org.apache.zookeeper.ClientCnxn - Session 0x0 for server null, unexpected error, closing socket connection and attempting reconnect
java.net.NoRouteToHostException: Host is unreachable
... 
(ONE HOUR LATER)
...
2019-04-17 17:32:39,224 4141095 [main-SendThread(172.24.57.9:2181)] INFO  org.apache.zookeeper.ClientCnxn - Opening socket connection to server 172.24.57.9/172.24.57.9:2181. Will not attempt to authenticate using SASL (unknown error)
2019-04-17 17:32:40,317 4142188 [core-14] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [0, 1, 2, 3], Current = [0, 1, 2, 3], PendingTasks = [], ToStart = [], ToStop = [].
2019-04-17 17:32:41,196 4143067 [main-SendThread(172.24.57.9:2181)] WARN  org.apache.zookeeper.ClientCnxn - Session 0x0 for server null, unexpected error, closing socket connection and attempting reconnect
java.net.NoRouteToHostException: Host is unreachable
```
```
(Case 2: zookeeperUri: zookeeper-0.zookeeper-headless.default.svc.cluster.local:2181)
----------------------------------------------------------
2019-04-17 14:50:15,403 3037022 [main-SendThread(zookeeper-0.zookeeper-headless.default.svc.cluster.local:2181)] INFO  org.apache.zookeeper.ClientCnxn - Unable to read additional data from server sessionid 0x10000eb977e0027, likely server has closed socket, closing socket connection and attempting reconnect
2019-04-17 14:50:15,405 3037024 [main-SendThread(zookeeper-0.zookeeper-headless.default.svc.cluster.local:2181)] INFO  org.apache.zookeeper.ClientCnxn - Unable to read additional data from server sessionid 0x10000eb977e0026, likely server has closed socket, closing socket connection and attempting reconnect
2019-04-17 14:50:15,505 3037124 [main-EventThread] INFO  o.a.b.z.ZooKeeperWatcherBase - ZooKeeper client is disconnected from zookeeper now, but it is OK unless we received EXPIRED event.
2019-04-17 14:50:15,506 3037125 [main-EventThread] INFO  org.apache.curator.ConnectionState - Connection string changed to: zookeeper-0.zookeeper-headless.default.svc.cluster.local:2181
2019-04-17 14:50:15,509 3037128 [main-EventThread] WARN  o.a.z.client.StaticHostProvider - No IP address found for server: zookeeper-0.zookeeper-headless.default.svc.cluster.local:2181
java.net.UnknownHostException: zookeeper-0.zookeeper-headless.default.svc.cluster.local: Name does not resolve
...
2019-04-17 14:51:12,744 3094363 [main-SendThread(zookeeper-0.zookeeper-headless.default.svc.cluster.local:2181)] INFO  org.apache.zookeeper.ClientCnxn - Client session timed out, have not heard from server in 11546ms for sessionid 0x10000eb977e0027, closing socket connection and attempting reconnect
2019-04-17 14:51:13,032 3094651 [main-SendThread(172.24.57.5:2181)] WARN  org.apache.zookeeper.ClientCnxn - Session 0x0 for server null, unexpected error, closing socket connection and attempting reconnect
java.net.NoRouteToHostException: Host is unreachable
...
2019-04-17 15:23:53,535 5055154 [main-SendThread(zookeeper-0.zookeeper-headless.default.svc.cluster.local:2181)] INFO  org.apache.zookeeper.ClientCnxn - Opening socket connection to server zookeeper-0.zookeeper-headless.default.svc.cluster.local/172.24.57.5:2181. Will not attempt to authenticate using SASL (unknown error)
2019-04-17 15:23:55,304 5056923 [main-SendThread(172.24.57.5:2181)] WARN  org.apache.zookeeper.ClientCnxn - Session 0x0 for server null, unexpected error, closing socket connection and attempting reconnect
java.net.NoRouteToHostException: Host is unreachable
```

_BONUS TRACK 2_: In the case for which the Controller restarts ungracefully, we receive a `SIGTERM` when trying to re-create the Curator client:
```
2019-04-18 08:43:31,066 235709 [Thread-2] INFO  io.pravega.controller.server.Main - Shutdown Hook Thread dump: Thread SIGTERM handler stackTrace: [java.lang.Object.wait(Native Method), java.lang.Thread.join(Thread.java:1252), java.lang.Thread.join(Thread.java:1326), java.lang.ApplicationShutdownHooks.runHooks(ApplicationShutdownHooks.java:107), java.lang.ApplicationShutdownHooks$1.run(ApplicationShutdownHooks.java:46), java.lang.Shutdown.runHooks(Shutdown.java:123), java.lang.Shutdown.sequence(Shutdown.java:167), java.lang.Shutdown.exit(Shutdown.java:212), java.lang.Terminator$1.handle(Terminator.java:52), sun.misc.Signal$1.run(Signal.java:212), java.lang.Thread.run(Thread.java:748)]
``` 
No clue about the relationship between the endpoint passed to Curator and the fact of getting a `SIGTERM`.

**Problem location**
Interaction between Curator and Zookeeper in Kubernetes cluster (DNS service?).

**Suggestions for an improvement**
Allow Curator to get the right IP of the new Zookeeper pod after it restarts.

**Comment by RaulGracia** [comment]
After extensive investigation and several attempts, I have some insights to make the interaction between Pravega and Zookeeper more stable in Kubernetes:
1) If we use the `headless` endpoint in the Pravega manifest, the Controller can restart gracefully (and we avoid the still mysterious `SIGTERM`)
2) In the Segment Store, we have identified 2 problems when a ZK pod restarts: i)  It resolves the ZK hostname only once when the Curator client is created, so it gets stuck when the ZK pod changes the IP; and ii) the Zookeeper client does not re-resolve the ZK endpoint. I have fixed the first problem using a custom `ZKClientFactory` in the Segment Store that ensures that we always use the hostname when creating a ZK client in Curator. The second issue is fixed by using `zookeeper-3.5.5` in Curator.

With these fixes, I executed the same experiment with all these fixes: A Pravega Cluster with 1 Controller, 1 SSS and 1 ZK pod. I used the Pravega benchmark to write some data:
```
./bin/pravega-benchmark -controller tcp://172.24.38.3:9090 -producers 1 -segments 1 -size 100 -events 10 -time 20 -stream test1
      51 records Writing,      10.2 records/sec,   0.00 MB/sec,    17.9 ms avg latency,   119.0 ms max latency
      50 records Writing,      10.0 records/sec,   0.00 MB/sec,     5.5 ms avg latency,     8.0 ms max latency
      50 records Writing,      10.0 records/sec,   0.00 MB/sec,     5.4 ms avg latency,     8.0 ms max latency
191 records Writing, 9.999 records/sec, 100 bytes record size, 0.00 MB/sec, 8.7 ms avg latency, 119.0 ms max latency, 5 ms 50th, 6 ms 75th, 7 ms 95th, 14 ms 99th, 14 ms 99.9th
```
Then, I deleted the only ZK pod and we see no Controller or SSS restarts:
```
kubectl get pods
NAME                                                    READY   STATUS    RESTARTS   AGE
benchmark-pod                                           1/1     Running   0          35h
lolling-coral-nfs-client-provisioner-76c99c5d4d-4ntvv   1/1     Running   0          36h
pravega-bookie-0                                        1/1     Running   0          37m
pravega-bookie-1                                        1/1     Running   0          37m
pravega-bookie-2                                        1/1     Running   0          37m
pravega-operator-5f94fc5c88-knp8s                       1/1     Running   0          36h
pravega-pravega-controller-67c56945f9-t4d2l             1/1     Running   0          37m
pravega-pravega-segmentstore-0                          1/1     Running   0          37m
zookeeper-0                                             1/1     Running   0          26m
zookeeper-operator-5cf75dcb56-zbwqj                     1/1     Running   0          36h
```
In particular, the SSS shows the following:
1) With the custom ZKClientFactory we use again the hostname (not the IP as before)
```
2019-04-25 09:34:29,643 611618 [main-EventThread] INFO  org.apache.zookeeper.ZooKeeper - Initiating client connection, connectString=zookeeper-0.zookeeper-headless.default.svc.cluster.local:2181 sessionTimeout=10000 watcher=org.apache.curator.ConnectionState@399f45b1
```

2) Zookeeper `3.5.5` changed the behavior, as we now see re-resolution attempts until the zookeeper IP is resolved:
```
2019-04-25 09:34:29,682 611657 [main-SendThread()] ERROR o.a.z.client.StaticHostProvider - Unable to resolve address: zookeeper-0.zookeeper-headless.default.svc.cluster.local:2181
java.net.UnknownHostException: zookeeper-0.zookeeper-headless.default.svc.cluster.local
...
2019-04-25 09:35:01,613 643588 [main-SendThread(zookeeper-0.zookeeper-headless.default.svc.cluster.local:2181)] ERROR o.a.z.client.StaticHostProvider - Unable to resolve address: zookeeper-0.zookeeper-headless.default.svc.cluster.local:2181
```
3) And then, the connection with the new pod is established (previous IP was `x.x.x.8`):
```
2019-04-25 09:35:02,781 644756 [ZKC-connect-executor-0-SendThread(zookeeper-0.zookeeper-headless.default.svc.cluster.local:2181)] INFO  org.apache.zookeeper.ClientCnxn - Opening socket connection to server zookeeper-0.zookeeper-headless.default.svc.cluster.local/172.24.38.10:2181. Will not attempt to authenticate using SASL (unknown error)
2019-04-25 09:35:02,782 644757 [ZKC-connect-executor-0-SendThread(zookeeper-0.zookeeper-headless.default.svc.cluster.local:2181)] INFO  org.apache.zookeeper.ClientCnxn - Socket connection established, initiating session, client: /172.24.38.8:41878, server: zookeeper-0.zookeeper-headless.default.svc.cluster.local/172.24.38.10:2181
2019-04-25 09:35:02,785 644760 [ZKC-connect-executor-0-SendThread(zookeeper-0.zookeeper-headless.default.svc.cluster.local:2181)] INFO  org.apache.zookeeper.ClientCnxn - Session establishment complete on server zookeeper-0.zookeeper-headless.default.svc.cluster.local/172.24.38.10:2181, sessionid = 0x100083502ef0001, negotiated timeout = 10000
2019-04-25 09:35:02,785 644760 [ZKC-connect-executor-0-EventThread] INFO  o.a.b.z.ZooKeeperWatcherBase - ZooKeeper client is connected now.
```
4) Once both the Controller and Segment Store reconnect to Zookeeper, the system is able to handle IO again (this was not possible before):
```
root@benchmark-pod:/tmp/pravega-benchmark/build/distributions/pravega-benchmark# ./bin/pravega-benchmark -controller tcp://172.24.38.3:9090 -producers 1 -segments 1 -size 100 -events 10 -time 20 -stream test5
      51 records Writing,      10.2 records/sec,   0.00 MB/sec,    26.0 ms avg latency,   106.0 ms max latency
      50 records Writing,      10.0 records/sec,   0.00 MB/sec,     4.5 ms avg latency,     8.0 ms max latency
      50 records Writing,      10.0 records/sec,   0.00 MB/sec,     5.2 ms avg latency,    16.0 ms max latency
192 records Writing, 10.015 records/sec, 100 bytes record size, 0.00 MB/sec, 10.5 ms avg latency, 106.0 ms max latency, 5 ms 50th, 5 ms 75th, 9 ms 95th, 81 ms 99th, 81 ms 99.9th
```

*Summary*:I think that there are 3 main action points to be done:
1) Use by default the `headless` ZK endpoint in the Pravega Operator manifest. This may not require code changes, but we need to follow this guideline in our tests.
2) I will send out a PR with the Zookeeper client factory for the Segment Store.
3) I will contact with the Curator community to understand if it is possible to get soon a version of Curator `4.0.x` or `4.1.x` that includes `zookeeper-3.5.5`, which is about to get released: https://github.com/apache/zookeeper/releases.

**Comment by RaulGracia** [comment]
I created a Jira issue in the Curator project asking if it is in the roadmap to release a version of Curator `4.0.x` and/or `4.1.x` using `zookeeper-3.5.5`: https://issues.apache.org/jira/browse/CURATOR-519
